harness/determined/core/_profiler.py

```diff
@@ -0,0 +1,396 @@
+        if self._collector:
+            self._collector.stop()
+        if self._shipper:
+            self._shipper.stop()
+
+
+class DummyProfilerContext(ProfilerContext):
+    def __init__(
+        self,
+    ) -> None:
+        pass
+
+    def on(self, interval: int = 1) -> None:
+        pass
+
+    def off(self) -> None:
+        pass
+
+    def close(self) -> None:
+        pass
+
+
+class _MetricGroupCollector(metaclass=abc.ABCMeta):
+    group: str
+
+    def __init__(self):
+        pass
+
+    @abc.abstractmethod
+    def collect(self) -> Dict[str, Any]:
+        pass
+
+
+class _Network(_MetricGroupCollector):
+    group = "network"
+
+    def __init__(self) -> None:
+        super().__init__()
+
+        # Set initial values for throughput calculations.
+        self._start_ts = time.time()
+        self._start_vals = psutil.net_io_counters()
+
+    def collect(self) -> Dict[str, Any]:
+        ts = time.time()
+        vals = psutil.net_io_counters()
+
+        sent_thru = (vals.bytes_sent - self._start_vals.bytes_sent) / (ts - self._start_ts)
+        recv_thru = (vals.bytes_recv - self._start_vals.bytes_recv) / (ts - self._start_ts)
+
+        self._start_ts, self._start_vals = ts, vals
+
+        return {
+            "net_throughput_sent": sent_thru,
+            "net_throughput_recv": recv_thru,
+        }
+
+
+class _Disk(_MetricGroupCollector):
+    group = "disk"
+
+    _disk_paths = ["/", constants.SHARED_FS_CONTAINER_PATH]
+
+    def __init__(self) -> None:
+        # Set initial values for throughput calculations.
+        self._start_ts = time.time()
+        self._start_vals = psutil.disk_io_counters()
+
+        # Initialize accessible disk paths.
+        self._paths = []
+        for path in self._disk_paths:
+            try:
+                psutil.disk_usage(path)
+                self._paths.append(path)
+            except Exception:
+                pass
+
+        super().__init__()
+
+    def collect(self) -> Dict[str, Any]:
+        ts = time.time()
+        vals = psutil.disk_io_counters()
+
+        read_thru = (vals.read_bytes - self._start_vals.read_bytes) / (ts - self._start_ts)
+        write_thru = (vals.write_bytes - self._start_vals.write_bytes) / (ts - self._start_ts)
+        iops = (vals.read_count + vals.write_count) - (
+            self._start_vals.read_count + self._start_vals.write_count
+        ) / (ts - self._start_ts)
+        self._start_ts, self._start_vals = ts, vals
+
+        metrics = {
+            "disk_iops": iops,
+            "disk_throughput_read": read_thru,
+            "disk_throughput_write": write_thru,
+        }
+
+        for path in self._paths:
+            disk_usage = psutil.disk_usage(path)
+            metrics.update({path: {"disk_util": disk_usage.percent}})
+        return metrics
+
+
+class _Memory(_MetricGroupCollector):
+    group = "memory"
+
+    def collect(self) -> Dict[str, Any]:
+        free_mem_bytes = psutil.virtual_memory().available
```

Comment:
emit more than just free memory; free memory is [not a very complete picture](https://www.linuxatemyram.com/) of memory usage.